J2735-2024 Part 9 of 9: Test Data Tools

iyourshaw opened the pull request:
# Part 9
## Includes:
* `jpo-asn-test-generator` - A Java command line tool for generating test messages documented here: [jpo-asn-test-generator/README](https://github.com/CDOT-CV/jpo-asn-pojos/blob/j2735-2024-part-9/jpo-asn-test-generator/README.md)
* `generate-test-data` - Folder contains example data and command line tools that were used to generate and verify the test messages for the POJOs, documented here: [generate-test-data/README](https://github.com/CDOT-CV/jpo-asn-pojos/blob/j2735-2024-part-9/generate-test-data/README.md)

## Tests
The test generator CLI tool includes [unit tests](https://github.com/CDOT-CV/jpo-asn-pojos/blob/j2735-2024-part-9/jpo-asn-test-generator/src/test/java/us/dot/its/jpo/asn/testgenerator/TestGenCliTest.java) that test it's ability to generate examples of every message type in the J2735/2024 specification.  Coverage report:

![image](https://github.com/user-attachments/assets/a5c25d27-dcd3-4c48-96b5-7651f14b8b0c)

iyourshaw committed:
Test generator

iyourshaw committed:
Merge branch 'j2735-2024-part-8' into j2735-2024-part-9

iyourshaw committed:
Update testgen instructions

iyourshaw committed:
Merge from part 8

iyourshaw committed:
Remove deserializers from all integers

iyourshaw committed:
Test deserializing large xml integer

iyourshaw committed:
Merge runtime updates

iyourshaw committed:
Handle deserializing JER format with bitstring value/length object

iyourshaw committed:
Bitstring serializer fix: serialize with variable format

iyourshaw committed:
Update MSCM json test

iyourshaw committed:
Updates to cli tool from mouse branch of asn1c

iyourshaw committed:
gitattributes for binary files

iyourshaw committed:
Add rawtypes annotations

iyourshaw committed:
Merge branch 'j2735-2024-part-9' into j2735-2024-part-9-var-bitstring-encoding

iyourshaw committed:
New jer converter.  Update tests with variable and extensible bitstrings for bsm, ic, sdsm, pvd.  Fix bitstring logic

iyourshaw committed:
Regenerate bitstrings with corrected hasExtensionMarker

iyourshaw committed:
Update tests for messages with extensible bitstrings

iyourshaw committed:
Add documentation for RootSeqeunceOfSerializer.  Fix TollPointMap.

iyourshaw committed:
Regenerate

iyourshaw committed:
Merge from develop

iyourshaw committed:
Fix tlsm test

iyourshaw committed:
roundtrip script

iyourshaw committed:
null tests

iyourshaw committed:
message frame test

iyourshaw committed:
RptRoadwayEventsTest

iyourshaw committed:
RptRoadwayEventsTests

iyourshaw committed:
Regenerate null types with deserializers

iyourshaw committed:
Tests for null types

iyourshaw committed:
MessageFrame tests

iyourshaw committed:
Tests for null choice types

iyourshaw committed:
named integers

iyourshaw committed:
Generate named integers

iyourshaw committed:
itis code tests

iyourshaw committed:
Coverage

iyourshaw committed:
Named integer test coverage

iyourshaw committed:
Remove unused

iyourshaw committed:
Named integer values from value assignments

iyourshaw committed:
Test for DSRCmsgID

iyourshaw committed:
Generator: Use spotless plugin to apply Google code format

iyourshaw committed:
Adding unit tests for test generator cli

iyourshaw committed:
Fix nested SEQUENCE types in PVD.VehicleStatus

iyourshaw committed:
Add ProbeVehicleData tests

iyourshaw committed:
Tests for generating all message types

iyourshaw committed:
Fix ints without constraints

iyourshaw committed:
merge from part 8

iyourshaw committed:
Test generator documentation

iyourshaw committed:
Roundtrip script docs

iyourshaw committed:
Code cleanup

iyourshaw committed:
Add examples to docs. Optional flags

iyourshaw committed:
test log config

iyourshaw committed:
usage/readme

iyourshaw committed:
Remove

iyourshaw committed:
Merge branch 'j2735-2024-part-8' into j2735-2024-part-9

iyourshaw committed:
edit

iyourshaw committed:
ignore IDE files

iyourshaw committed:
ignore IDE files

iyourshaw committed:
Merge branch 'j2735-2024-part-8' into j2735-2024-part-9

iyourshaw committed:
Add test-generator coverage reports to CI.  Make the coverage comment 'update' instead of create a new comment every time.

iyourshaw committed:
Fix ci duplicate name

iyourshaw committed:
Separate task for each coverage report

iyourshaw committed:
coverate reports

iyourshaw committed:
typo in readme

iyourshaw committed:
typos in readme

iyourshaw committed:
rawtypes annotations

iyourshaw committed:
Merge branch 'j2735-2024-part-8' into j2735-2024-part-9

iyourshaw committed:
Merge branch 'develop' into j2735-2024-part-9

payneBrandon commented on .github/workflows/test.yml:
```diff
@@ -19,22 +19,46 @@ jobs:
       uses: gradle/gradle-build-action@v2
     
-    - name: Run Tests with Coverage
+    - name: Run Tests with Coverage (jpo-asn-j2735-2024)
```
question: with the latest update to include both j2735 and runtime tests in the same run, should this reference both for clarity?

iyourshaw replied:
The github workflow uses gradle to build and post coverage comments to the PRs.  Maven jacoco is configured to aggregate, but the gradle version of jacoco doesn't aggregate: it only shows the coverage for one project.  Since this PR adds the test-generator project, it adds another task to display coverage for it, but that label is accurate as is, since the runtime project coverage isn't shown in those comments.

payneBrandon replied:
That's an unfortunate limitation in gradle then. Do we need to also have coverage report for the runtime portion?

iyourshaw replied:
It turns that that Gradle uses a separate plugin to do coverage aggregation.  I added that in and changed the workflow and readme instructions to match.  (You need to run 'gradle check' instead of 'gradle test' for it to work).  I tested that it works in a separate PR here: https://github.com/CDOT-CV/jpo-asn-pojos/pull/25

payneBrandon reviewed (commented):
There's a couple empty files that won't let me  comment in, the generate-test-data/map/.json and .xml. Are these empty files needed?

payneBrandon reviewed (commented):
There's a couple empty files that won't let me  comment in, the generate-test-data/map/.json and .xml. Are these empty files needed?

payneBrandon commented on generate-test-data/examples/2024/map/map-node-attribute-set-gen.xml:
```diff
@@ -0,0 +1,2735 @@
+<MapData>
```
question: we've got a mix of minified and expanded files here should we minimize them all to reduce space?

iyourshaw replied:
The minified files are the raw output of the generator tools.   Minified XML is considered to be "canonical" and is what the asn1c produces, too.  But the java POJOs ignore formatting and are capable of using either.  In fact it is a goal for them to be robust if they receive xml or json with different formatting.  Some of the examples are pretty-printed for that reason, and for others it makes them more human readable to illustrate some tricky parts of the XER and JER formats.  I'd prefer to keep as-is for those reasons, since the space savings would be minimal.

payneBrandon replied:
Ok, can we update the documentation to reflect this decision? It's a bit jarring to open files that are minified next to ones that are readable without the explanation.

iyourshaw replied:
Added a readme to the examples folder documenting that:
https://github.com/CDOT-CV/jpo-asn-pojos/blob/j2735-2024-part-9/generate-test-data/examples/README.md

iyourshaw committed:
delete empty files

iyourshaw commented:
> There's a couple empty files that won't let me comment in, the generate-test-data/map/.json and .xml. Are these empty files needed?

Deleted the empty files

payneBrandon commented on generate-test-data/examples/2024/pdc/pdc_rt.json:
```diff
@@ -0,0 +1,202 @@
+[
```
similar question: should this be minified?

iyourshaw commented:
Note on why the `jpo-asn-test-generator` project doesn't use Maven:

This is a CLI tool that builds with Gradle that is intended to be built and used locally.  It uses 'includeBuild' within `jpo-asn-test-generator/settings.gradle` to include the `jpo-asn-j2735-2024` and `jpo-asn-runtime` projects that it depends on, without having to have a "parent" project that all three projects depend on.

The Maven build that is available for `jpo-asn-runtime` and `jpo-asn-j2735-2024` is designed to be used to import those libraries into the ODE.  There is a `pom.xml` at the root level of the repo that includes both libraries.  The root-level pom doesn't include the `jpo-asn-test-generator` project, since it is intended to be built and used separately as a CLI tool, not imported as a library.

payneBrandon commented on jpo-asn-j2735-2024/src/main/java/us/dot/its/jpo/asn/j2735/r2024/AddGrpB/Angle.java:
```diff
@@ -0,0 +1,39 @@
+/*==============================================================================
```
question: the j2735 and runtime changes weren't mentioned in the description, are these related to a particular set of message types?

iyourshaw replied:
All the changes under jpo-asn-j2735-2024 and jpo-asn-runtime are cumulative from the Part 8 PR.  Sorry for the confusion caused by releasing these at the same time..

payneBrandon reviewed (changes requested):
Jsut a couple questions so far

payneBrandon commented on jpo-asn-runtime/src/main/java/us/dot/its/jpo/asn/runtime/serialization/OpenTypeDeserializer.java:
```diff
@@ -9,6 +9,7 @@
 import com.fasterxml.jackson.databind.JsonNode;
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.fasterxml.jackson.databind.deser.std.StdDeserializer;
+import com.fasterxml.jackson.databind.node.ArrayNode;
```
question: this doesn't appear to be used, do we need it?

iyourshaw replied:
Good catch, removed the unused import.

payneBrandon commented on jpo-asn-runtime/src/main/java/us/dot/its/jpo/asn/runtime/types/Asn1Integer.java:
```diff
@@ -1,66 +1,83 @@
+    return Long.toString(value);
+  }
+
+  public Optional<String> name() {
```
question: what's this used for if we always return Optional.empty?

iyourshaw replied:
It is the default in the base class for most integer types that don't have named values, if you ask for the name, it returns nothing.  But it is overridden for ints with named values to return the name if there is one.

iyourshaw committed:
Use jacoco aggregation plugin

iyourshaw committed:
Change whitespace to test runtime coverage

iyourshaw committed:
update docs

iyourshaw committed:
Fix artifacts upload

iyourshaw committed:
comment label

iyourshaw committed:
Merge pull request #25 from CDOT-CV/j2735-2024-part-9-coverage-aggregation

Test using jacoco coverage aggregation plugin

iyourshaw committed:
Add runtime to task names

iyourshaw committed:
Add runtime to task names

iyourshaw committed:
Fix

iyourshaw committed:
Merge branch 'j2735-2024-part-8' into j2735-2024-part-9

mcook42 committed:
chore: use update-index --chmod=+x to track all gradlew scripts as executable

mcook42 committed:
refactor: update value assignments to use setValue method

Replaced direct value assignments with the setValue method across multiple classes for better encapsulation and consistency. Also updated access modifiers in Asn1CharacterString to enforce proper encapsulation.

mcook42 committed:
refactor: encapsulate fields

Changed protected and package-private fields to private and provided getter methods to enhance encapsulation and maintainability. Updated references to these fields to use the new getter methods, ensuring consistency across the codebase.

iyourshaw committed:
Merge branch 'j2735-2024-part-8' into j2735-2024-part-9

iyourshaw committed:
update testgen jar

mcook42 committed:
Merge branch 'j2735-2024-part-8' of github.com:CDOT-CV/jpo-asn-pojos into mcook42/suggest/j2735-2024-part-8-part2

mcook42 committed:
refactor: Asn1RelativeOID and RelativeRoadAuthorityID

Refactored Asn1RelativeOID to remove the no-arg constructor and make the value field final, improving immutability. Updated RelativeRoadAuthorityID to include a JsonCreator annotation while removing the no-arg constructor for consistency. These changes enhance object safety and simplify usage.

mcook42 committed:
fix: correct constructor parameter in BooleanDeserializer

Updated the constructor to accept the provided value type instead of a hardcoded class. Also removed an unnecessary suppression warning to improve code clarity.

mcook42 committed:
refactor: change constructors to protected visibility

Updated constructors for Asn1Bitstring and Asn1CharacterString to use protected visibility instead of public. This change limits direct instantiation outside the package, ensuring better encapsulation and supporting inheritance where needed.

mcook42 committed:
refactor: simplify Boolean deserializers and remove redundancies

Removed custom inner deserializer classes and replaced them with a generalized `BooleanDeserializer` implementation. This reduces code duplication across multiple classes and improves maintainability by centralizing deserialization logic.

mcook42 committed:
refactor: centralize BitString deserialization logic

Removed custom deserializers from individual classes and centralized the logic into a unified `BitStringDeserializer`. This improves maintainability and simplifies the codebase by handling deserialization contextually in a single location.

mcook42 committed:
fix: handle null BeanProperty in BooleanDeserializer

Ensure proper handling of null BeanProperty in the `createContextual` method. This fix prevents potential null pointer exceptions by using the contextual type when the property is not provided.

mcook42 committed:
Merge branch 'j2735-2024-part-9' of github.com:CDOT-CV/jpo-asn-pojos into mcook42/suggest/j2735-2024-part-8-part2

mcook42 committed:
refactor: change populateRandom to return the instance

Modified all populateRandom methods to return the given instance, ensuring consistency across all generator classes. This allows us to respect encapsulation

mcook42 commented on .github/workflows/test.yml:
```diff
@@ -19,22 +19,46 @@ jobs:
+    - name: Run Tests with Coverage (jpo-asn-test-generator)
+      working-directory: ./jpo-asn-test-generator
+      run: |
+        chmod +x gradlew
```
suggestion(non-blocking): track `gradlew` as executable in git so that we don't need to `chmod +x`. I've done this already over in https://github.com/CDOT-CV/jpo-asn-pojos/pull/26

mcook42 commented on generate-test-data/README.md:
````diff
@@ -105,14 +98,30 @@ This converter consumes and produces **CANONICAL** XER which is the same dialect
+Example. To convert the above SDSM XER to JER:
+
+```bash
+./converter-all -p SensorDataSharingMessage -ixer -ojer sdsm1.xml > sdsm1.json
````
question(blocking): Is this documentation still relevant? I don't see a `converter-all` binary included, and I don't see any documentation discussing how to generate the `converter-all` binary

iyourshaw replied:
That is an old example, updated it to refer to "converter-debug"

iyourshaw committed:
Merge branch 'develop' into j2735-2024-part-9

iyourshaw committed:
Update example

iyourshaw committed:
Use Jackson exceptions, remove hardcoded example

iyourshaw committed:
spotless plugin/google format

iyourshaw committed:
spotlessApply google format

iyourshaw committed:
Revert format changes to generator

iyourshaw committed:
Fix pojo spotless config.  Apply spotless to generator files

mcook42 committed:
Merge pull request #27 from CDOT-CV/iyourshaw/suggest/j2735-2024-part-8-part2

Suggestions for "Suggestion: Refactor Deserializers"

iyourshaw committed:
Merge pull request #26 from CDOT-CV/mcook42/suggest/j2735-2024-part-8-part2

Suggestion: Refactor Deserializers

mcook42 commented on jpo-asn-j2735-2024/README.md:
````diff
@@ -246,14 +246,12 @@ automated tools to aid in creating and verifying the test messages. Those tools
 
 ```bash
-./gradle clean test
+./gradle clean check
````
```suggestion
./gradlew clean check

```

iyourshaw committed:
Update jpo-asn-j2735-2024/README.md

Co-authored-by: Matt Cook <mattheworion.cook@gmail.com>

payneBrandon commented on generate-test-data/README.md:
```diff
@@ -105,14 +98,30 @@ This converter consumes and produces **CANONICAL** XER which is the same dialect
 
+### converter-new-jer
+
+A converter generated from the `vlm_master` branch of the [mouse07410 fork](https://github.com/mouse07410/asn1c) of asn1c, which contains updates to the JER implementation that aren't present in the USDOT fork.  Specifically, it contains changes to the implementation of variable length bitstring encodings in JER.
```
question: is this converter documentation the same as the debug version above? IE are parameters expected to be the same?

iyourshaw replied:
Yes, added an example of using it to illustrates that.

payneBrandon commented on jpo-asn-j2735-2024/src/main/java/us/dot/its/jpo/asn/j2735/r2024/J2540ITIS/ITIStext.java:
```diff
@@ -34,6 +34,6 @@ public ITIStext() {
   public ITIStext(String value) {
     this();
-    this.value = value;
+    this.setValue(value);
```
question: As I'm going through the changed files, I see that we have duplicate `ITIStext.java` classes with the only difference being the namespace. Is this due to some asn structuring, or is there merit in merging the redundant classes?

iyourshaw replied:
It is a literal translation of the module structure of the asn.1.  There are modules named both ITIS and J2540ITIS that contain some types with the same names.  I think it is best to preserve it as is to reflect what is in the standard rather than alter it manually.

iyourshaw committed:
converter-new-jer example

iyourshaw committed:
Merge branch 'j2735-2024-part-9' of https://github.com/CDOT-CV/jpo-asn-pojos into j2735-2024-part-9

iyourshaw committed:
Document minified & pretty-printed examples

iyourshaw committed:
Document minified & pretty-printed examples

payneBrandon reviewed (approved):
lgtm!
